### PR TITLE
Posix x64 ABI fixes 2

### PR DESCRIPTION
**Pull request by SSoulaimane**
Corrects DMD's nonadherence to the C ABI in Posix x64.
Fixes issues [13957](https://issues.dlang.org/show_bug.cgi?id=13957) and [5570](https://issues.dlang.org/show_bug.cgi?id=5570).

This is a reboot of #9013 with less boilerplate. It solves all cases of structs containing integral and/or floats, but doesn't touch 80bit real and SIMD vectors, these are planned to be followups in separate PRs.

Windows x64 is planned next.

### CODE REVIEW

File: src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -1251,6 +1334,220 @@ version (MARS)
...
+                else
+                    return 0;
+                if (targ2)
+                    ty2 = targ2.Tty;
+                break;
+            }
+            else if (!(t.Ttag.Sstruct.Sflags & STRnotpod))
+            {
+                // windows only, return POD of 1, 2, 4, or 8 bytes on EAX(:EDX)
+                if (0 == (config.exe & (EX_WIN64 | EX_WIN32)))
```
Comment (thewilsonator) [inline src/dmd/backend/cod3.d]: `if (!(config.exe & (EX_WIN64 | EX_WIN32))`

File: src/dmd/backend/cod1.d [hunk 2]
```diff
@@ -3117,6 +3153,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+                if (tyfloating(tyn))
+                {
+                    *arg1type = sz <= 4 ? tstypes[TYfloat] : tstypes[TYdouble];
+                    if (sz > 8)
+                        *arg2type = (sz - 8) <= 4 ? tstypes[TYfloat] : tstypes[TYdouble];
+                }
+            }
+        }
+    }
+    else if (tybasic(ty) == TYstruct)
```
Comment (thewilsonator) [inline src/dmd/backend/cod1.d]: ?

File: src/dmd/backend/cod1.d [hunk 3]
```diff
@@ -3117,6 +3153,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
     return 1;
 }
 
+/***************************************
+ * Finds replacemnt types for register passing of aggregates.
+ */
+void argtypes(type *t, type **arg1type, type **arg2type)
```
Comment (thewilsonator) [inline src/dmd/backend/cod1.d]: can `arg1type` and `arg2type` be `ref type*` instead? 
Also please write `type* arg` not `type *arg`, as this is not C.

File: src/dmd/backend/cod1.d (hunk 2 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: Placeholder for potential moving the some of the code in `cgelem.d:elstruct()` here in the future.

File: src/dmd/backend/cod1.d [hunk 4]
```diff
@@ -3019,10 +3030,27 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
             t = null;
             ty = TYnptr;
         }
+        else if (!tyaggregate(t.Tty))
+        {
+            // it happens with arrays sometimes, `t` is set to the element type
```
Comment (thewilsonator) [inline src/dmd/backend/cod1.d]: when does it happen with arrays?

File: src/dmd/backend/cod1.d (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/backend/cod1.d]: Please, put a comment saying that then.

File: src/dmd/backend/cod1.d (hunk 3 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: OK.

File: src/dmd/backend/cod1.d (hunk 4 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: I don't remember. I'll recheck;

Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: I couldn't reproduce it, I removed it.

**Comment by RazvanN7** [comment]
cc @WalterBright

**Comment by WalterBright** [comment]
With 800+ lines of changes in the code generator, I need some time to review this.

**Comment by wilzbach** [comment]
> With 800+ lines of changes in the code generator, I need some time to review this.

CC @WalterBright

**Comment by Geod24** [comment]
Since you already have a very good commit separation, with proper messages, can you gradually submit smaller PRs ? The number of people able to review this is very small, and they are usually overloaded, so being able to gradually review/merge might unstuck this.

**Comment by SSoulaimane** [comment]
@Geod24 maybe that could move things faster.

**Comment by WalterBright** [comment]
Some exposition on what the top level strategy being used here would be nice, as it's hard to deduce from the bottom level implementation.

File: src/dmd/backend/cod1.d [hunk 5]
```diff
@@ -3118,6 +3149,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+void argtypes(type* t, ref type* arg1type, ref type* arg2type)
+{
+    if (!t) return;
+
+    tym_t ty = t.Tty;
+
+    if (!tyaggregate(ty))
+        return;
+
+    arg1type = arg2type = null;
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: Clearly, these should be `out` parameters, not `ref` parameters.

File: src/dmd/backend/cod1.d [hunk 6]
```diff
@@ -3118,6 +3149,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+
+    if (tybasic(ty) == TYarray)
+    {
+        size_t sz = cast(size_t) type_size(t);
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: `type_size` returns `targ_size_t`, why cast it?

File: src/dmd/backend/cod3.d [hunk 7]
```diff
@@ -1111,7 +1111,37 @@ static if (NTEXCEPTIONS)
...
+                lreg = XMM0;
+                mreg = XMM1;
+            }
+            else
+            {
+                lreg = mask(reg1) & mLSW ? reg1 : AX;
+                mreg = mask(reg2) & mMSW ? reg2 : DX;
+            }
+            if (reg1 != NOREG)
+                retregs = (mask(lreg) | mask(mreg)) & ~mask(NOREG);
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: What is the purpose of calling allocretregs on line 1120 if it gets redone here?

File: src/dmd/backend/cod3.d [hunk 8]
```diff
@@ -1111,7 +1111,37 @@ static if (NTEXCEPTIONS)
...
+            reg1 = reg2 = NOREG;
+            if (config.exe == EX_WIN64) // broken
+                retregs = regmask(e.Ety, funcsym_p.ty());
+            else
+            {
+                retregs = allocretregs(e.Ety, e.ET, funcsym_p.ty(), &reg1, &reg2);
+                assert(reg1 != NOREG || !retregs);
+            }
+
+            lreg = mreg = NOREG;
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: distribute these defaults into each of the { } clauses below.

File: src/dmd/backend/cod3.d [hunk 9]
```diff
@@ -1142,6 +1172,59 @@ static if (NTEXCEPTIONS)
             {
                 gencodelem(cdb,e,&retregs,true);
             }
+
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: I do not know what the following lines of code do.

File: src/dmd/backend/cod3.d [hunk 10]
```diff
@@ -1142,6 +1172,59 @@ static if (NTEXCEPTIONS)
...
+                else
+                {
+                    assert(reg1 == AX && reg2 == DX);
+                    regm_t pretregs = mask(reg1) | mask(reg2);
+                    fixresult_complex87(cdb, e, retregs, &pretregs);
+                }
+            }
+            else if (reg2 == NOREG)
+                assert(lreg == reg1);
+            else for (int v = 0; v < 2; v++)
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: Don't put `else for` on one line, very confusing.

File: src/dmd/backend/cod3.d [hunk 11]
```diff
@@ -1142,6 +1172,59 @@ static if (NTEXCEPTIONS)
...
+            {
+                if (v ^ (reg1 != mreg))
+                    genmovreg(cdb, reg1, lreg);
+                else
+                    genmovreg(cdb, reg2, mreg);
+            }
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: This loop just seems weird. So much simpler to replace it with:
```
if (reg1 == mreg)
{
    genmovreg(cdb, reg2, mreg);
    genmovreg(cdb, reg1, lreg);
}
else
{
    genmovreg(cdb, reg1, lreg);
    genmovreg(cdb, reg2, mreg);
}

File: src/dmd/backend/cod3.d [hunk 12]
```diff
@@ -1251,6 +1334,220 @@ version (MARS)
...
+ * Returns:
+ *    a bit mask of return registers.
+ *    0 if function returns on the stack or returns void.
+ */
+regm_t allocretregs(tym_t ty, type *t, tym_t tyf, reg_t *reg1, reg_t *reg2)
+{
+    tym_t ty1 = ty;
+    tym_t ty2 = TYMAX;
+
+    *reg1 = *reg2 = NOREG;
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: Use `out` parameters for these two.

File: src/dmd/backend/cod1.d [hunk 13]
```diff
@@ -3118,6 +3149,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+        if (sz == 0)
+            return;
+
+        if ((I32 || config.exe == EX_WIN64) && (sz & (sz - 1)))  // power of 2
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: that means NOT a power of 2

File: src/dmd/backend/cod1.d [hunk 14]
```diff
@@ -3118,6 +3149,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+        {
+            type** argtype = &arg1type;
+            size_t argsz = sz < REGSIZE ? sz : REGSIZE;
+            foreach (v; 0 .. (sz > REGSIZE) + 1)
+            {
+                *argtype = argsz == 1 ? tstypes[TYchar]
+                         : argsz == 2 ? tstypes[TYshort]
+                         : argsz <= 4 ? tstypes[TYlong]
+                         : tstypes[TYllong];
+                argtype = &arg2type;
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: Should use a nested function for this, rather that the hoary old C method.

File: src/dmd/backend/cod1.d [hunk 15]
```diff
@@ -3118,6 +3149,108 @@ int FuncParamRegs_alloc(ref FuncParamRegs fpr, type* t, tym_t ty, reg_t* preg1,
...
+            {
+                if (type_size(tn) == sz) // array(s) of size 1
+                {
+                    arg1type = tn.Ttag.Sstruct.Sarg1type;
+                    arg2type = tn.Ttag.Sstruct.Sarg2type;
+                    return;
+                }
+
+                type* t1 = tn.Ttag.Sstruct.Sarg1type;
+                if (t1)
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: Merge previous two lines.

**Comment by WalterBright** [comment]
Rebased it.

**Comment by WalterBright** [comment]
Actually, I didn't rebase it. I just fixed the merge conflicts. How do I rebase someone else's PR?

**Comment by NilsLankila** [comment]
just like one of yours. add his repo as alternative origin, create a local branch from his repo, checkout then `git rebase -i master`, and then push force in his repo, which is allowed unless a specific setting has been uncheck.

**Comment by WalterBright** [comment]
@NilsLankila can you please rephrase that as specific git commands? Or, just do it, since I fixed the merge conflicts it should work.

**Comment by WalterBright** [comment]
Or maybe I can just take this PR over instead of force pushing in his repo?

**Comment by NilsLankila** [comment]
I'll do it. then. My workflow is partially based on git gui so it's faster for me to do it.

**Comment by WalterBright** [comment]
Thank you.

**Comment by NilsLankila** [comment]
There's been a little accident but it's good now apparently. For some reasons I had to fix the conflicts a second time. It's really not clear why. Usually when conflicts are already fixed you just `do git rebase -i master`, `ESC wq` to quit vi, and it's good. :confused:

**Comment by NilsLankila** [comment]
This PR was too big anyway. I'd suggest to pick commits one by one each time new PR, at least for those who looked ok to you.

**Comment by WalterBright** [comment]
I know it's too big and complex for one PR. That's why it's sat for so long :-(

**Comment by WalterBright** [comment]
Looks like it might be hanging on the 64 bit Posix tests.

**Comment by WalterBright** [comment]
@NilsLankila I'll resume work on this tomorrow. In the meantime, if you can get it to pass, feel free to turn on automerge.

**Review by WalterBright — APPROVED** [review]

**Comment by NilsLankila** [comment]
For now I have reset the feature branch to the state it was before the bad rebase. What happens is that the conflicts solved online works but when you rebase on master there are new conflicts, but **fewer** than there should be. This a pathological case.

**Comment by NilsLankila** [comment]
Rebasing tested in https://github.com/dlang/dmd/pull/11657. I still have a backup branch (representing the state just after that the conflicts were solved online) in case of.

**Comment by WalterBright** [comment]
Thanks @SSoulaimane and @NilsLankila !

**Comment by WalterBright** [comment]
@NilsLankila if there are other open @SSoulaimane PRs that could use rebasing, can you take care of those too, please?

**Comment by kinke** [comment]
Oh wow, I didn't think this day would come. Thanks @SSoulaimane, I hope you might consider contributing again sometime, a guy with your skills is hard to come by and always welcome. :)

**Comment by NilsLankila** [comment]
I'm gonna give a try in a few days and supposing not too much backedn knowledge ir required, unfortunately for now i'm indisposed with horrible throat aches :face_with_thermometer:

**Comment by dkorpel** [comment]
This PR introduced a regression: 
https://issues.dlang.org/show_bug.cgi?id=21328

**Pull request merged**